File: Python/Three_Sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        N = len(nums)
        result = []
        nums.sort() # Sort the array in-place
        
        # Fix the first number in triplet
        for i in range(N):
            # We initialise two pointers to search for the other two numbers 
            p1 = i+1
            p2 = N-1

            # If the first number was already seen before, skip it
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            # As long as the left is before right, continue search
            while p1 < p2:
                
                s = nums[i] + nums[p1] + nums[p2]
    
                if s == 0:
                    # We found a triplet so record it and change indices
                    result.append([nums[i], nums[p1], nums[p2]])
                    p1 += 1
                    p2 -= 1
                    
                    # Change left and right pointers until duplicates are skipped
                    while p1 < p2 and nums[p1] == nums[p1-1]:
                        p1 += 1
                    while p1 < p2 and nums[p2] == nums[p2+1]:
                        p2 -= 1
                    
                elif s < 0:
                    p1 += 1
                else:
                    p2 -= 1
                
        return result

File: C++/Three_Sum.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int N = nums.size();
        vector<vector<int>> result;

        //Sort the array in-place
        sort(nums.begin(), nums.end());
        
        // Fix the first number in triplet
        for (int i = 0; i < N; i++)
        {
            // We initialise two pointers to search for the other two numbers 
            int p1 = i+1;
            int p2 = N-1;
            
            // If the first number was already seen before, skip it
            if (i > 0 && nums[i] == nums[i-1]){ continue; }
            

            // As long as the left is before right, continue search
            while (p1 < p2)
            {
                int s = nums[i] + nums[p1] + nums[p2];
                
                if (s == 0)
                {
                    // We found a triplet so record it and change indices
                    vector<int> triplet = {nums[i], nums[p1], nums[p2]};
                    result.push_back(triplet);
                    p1++;
                    p2--;
                    

                    // Change left and right pointers until duplicates are skipped
                    while (p1 < p2 && nums[p1] == nums[p1-1])
                    {
                        p1++;
                    }
                    
                    while (p1 < p2 && nums[p2] == nums[p2+1])
                    {
                        p2--;
                    }
                }
                else if (s < 0)
                {
                    p1++;
                }
                else
                {
                    p2--;
                }
            }
            
        }
        
        return result;
    }
};